### ISSUE TITLE
Inconsistent results for `require: false` + `where` for LEFT OUTER JOIN across PostgreSQL and SQLite due to subquery vs JOIN-only and weird null handling

### ISSUE THREAD

**Issue by cirosantilli** [sequelize/sequelize#13809]
Also at: https://stackoverflow.com/questions/31091420/nodejs-sequelize-include-where-required-false/70446778#70446778

Main questions: is behavior intended (don't think so because SQLite and Postgres behave differently), and is there a better workaround than super many to many as shown on the SO answer?

Consider the following example where we want to:

- list all tags

- in addition, determine if each tag has been assigned to a selected animal

  Our where condition on the animal then is a simple "check if the animal is a specific animal", and can only return 0 or 1 entries therefore. But the same would be valid with any other more complex condition.

## Issue Creation Checklist

[X] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

### SSCCE

```
#!/usr/bin/env node
const assert = require('assert')
const path = require('path')
const { DataTypes, Sequelize } = require('sequelize')
let sequelize
if (process.argv[2] === 'p') {
  sequelize = new Sequelize('tmp', undefined, undefined, {
    dialect: 'postgres',
    host: '/var/run/postgresql',
  })
} else {
  sequelize = new Sequelize({
    dialect: 'sqlite',
    storage: 'tmp.sqlite'
  })
}
;(async () => {
const Animal = sequelize.define('Animal', {
  species: { type: DataTypes.STRING },
})
const Tag = sequelize.define('Tag', {
  name: { type: DataTypes.STRING },
})
Animal.belongsToMany(Tag, { through: 'AnimalTag' })
Tag.belongsToMany(Animal, { through: 'AnimalTag' })
await sequelize.sync({force: true})
const animal0 = await Animal.create({ species: 'bat' })
const animal1 = await Animal.create({ species: 'ant' })
const animal2 = await Animal.create({ species: 'dog' })
const tag0 = await Tag.create({ name: 'mammal' })
const tag1 = await Tag.create({ name: 'flying' })
const tag2 = await Tag.create({ name: 'aquatic' })
await animal0.addTag(tag0)
await animal0.addTag(tag1)
await animal2.addTag(tag0)
const animals = [animal0, animal1, animal2]
for (let animal of animals) {
  let rows
  rows = await Tag.findAll({
    include: [{
      model: Animal,
      where: { id: animal.id },
      // No effect.
      //on: { id: animal.id },
      through: {
        // Same as putting the where outside of through.
        //where: { AnimalId: animal.id },
      },
      required: false,
    }],
    order: [['name', 'ASC']],
  })
  console.error(animal.species);
  console.error(rows.map(row => { return {
    name: row.name,
    animals: row.Animals.map(animal => animal.species),
  } }))
}
})().finally(() => { return sequelize.close() })
```

package.json

```
{
  "name": "tmp",
  "private": true,
  "version": "1.0.0",
  "dependencies": {
    "pg": "8.5.1",
    "pg-hstore": "2.3.3",
    "sequelize": "6.5.1",
    "sqlite3": "5.0.2"
  }
}
```

PostgreSQL 13.5.

### What do you expect to happen?

On both SQLite and PostgreSQL:

```
bat
[
  { name: 'aquatic', animals: [] },
  { name: 'flying', animals: [ 'bat' ] },
  { name: 'mammal', animals: [ 'bat' ] }
]

ant
[
  { name: 'aquatic', animals: [] },
  { name: 'flying', animals: [] },
  { name: 'mammal', animals: [] }
]

dog
[
  { name: 'aquatic', animals: [] },
  { name: 'flying', animals: [] },
  { name: 'mammal', animals: [ 'dog' ] }
]
```

### What is actually happening?

PostgreSQL gives me what I wanted, SQLite gives:

```
bat
[
  { name: 'aquatic', animals: [] },
  { name: 'flying', animals: [ 'bat' ] },
  { name: 'mammal', animals: [ 'bat', null ] }
]

ant
[
  { name: 'aquatic', animals: [] },
  { name: 'flying', animals: [] },
  { name: 'mammal', animals: [] }
]

dog
[
  { name: 'aquatic', animals: [] },
  { name: 'flying', animals: [] },
  { name: 'mammal', animals: [] }
]
```

I don't mind the null that much, I can work around that.

But dog not being a mammal, that I cannot work around.

### Underlying queries

PostgreSQL has queries of type:

```
SELECT
  "Tag"."name",
  "Animals"."species" AS "Animals.species",
FROM
  "Tags" AS "Tag"
  LEFT OUTER JOIN (
    "AnimalTag" AS "Animals->AnimalTag"
    INNER JOIN "Animals" AS "Animals" ON "Animals"."id" = "Animals->AnimalTag"."AnimalId"
  ) ON "Tag"."id" = "Animals->AnimalTag"."TagId"
  AND "Animals"."id" = 1
ORDER BY
  "Tag"."name" ASC;
```

SQLite has queries of type:

```
SELECT
 `Tag`.`name`,
 `Animals`.`species` AS `Animals.species`
FROM `Tags` AS `Tag`
LEFT OUTER JOIN `AnimalTag` AS `Animals->AnimalTag`
  ON `Tag`.`id` = `Animals->AnimalTag`.`TagId`
LEFT OUTER JOIN `Animals` AS `Animals`
  ON `Animals`.`id` = `Animals->AnimalTag`.`AnimalId`
  AND `Animals`.`id`= 3
ORDER BY
 `Tag`.`name` ASC;
```

So only the PostgreSQL query is the desired one because of the way it uses the subquery so that the:

```
ON "Tag"."id" = "Animals->AnimalTag"."TagId"
  AND "Animals"."id" = 1
```

gets applied only to the outer `LEFT OUTER JOIN`.

The SQLite query produces undesired results, running it manually for each animal we get outputs:

```
bat

aquatic|
flying|bat
mammal|bat
mammal|

ant

aquatic|
flying|
mammal|
mammal|

dot

aquatic|
flying|
mammal|
mammal|dog
```

so what seems to happen is that the animal first result is null, then sequelize ignores it on the return, which would explain why we saw:

```
  { name: 'mammal', animals: [ 'bat', null ] }
```

for bat but:

```
  { name: 'mammal', animals: [] }
```

for dog.

The desired statement would have the `AND` on the first `LEFT OUTER JOIN`:

```
SELECT
 `Tag`.`name`,
 `Animals`.`species` AS `Animals.species`
FROM `Tags` AS `Tag`
LEFT OUTER JOIN `AnimalTag` AS `Animals->AnimalTag`
  ON `Tag`.`id` = `Animals->AnimalTag`.`TagId`
  AND `Animals->AnimalTag`.`AnimalId`= 1
LEFT OUTER JOIN `Animals` AS `Animals`
  ON `Animals`.`id` = `Animals->AnimalTag`.`AnimalId`
ORDER BY
 `Tag`.`name` ASC;
```

### Environment

- Sequelize version: 5.0.2 
- Node.js version: v14.17.0 

### How does this problem relate to dialects?

- [ ] I think this problem happens regardless of the dialect.
- [X] I think this problem happens only for the following dialect(s): SQLite
- [ ] I don't know, I was using PUT-YOUR-DIALECT-HERE, with connector library version XXX and database version XXX

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [X] Yes, I have the time but I don't know how to start, I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by WikiRik** [comment]
Does this happen on the latest release as well? v6.13.0 as of writing

**Comment by cirosantilli** [comment]
Thanks for looking into this, it is fixed on v6.13.0, sorry I didn't test with a newer version when reporting earlier :-)

**Comment by WikiRik** [comment]
No worries, good to hear that the issue is fixed!
